### PR TITLE
Add pull to refresh on all domains screen

### LINKED ISSUE #19562: 🤖 Add pull-to-refresh on All Domains screen
Add a pull-to-refresh behavior to the all domains screen. At a minimum, this should be implemented on the error screen after the fetch fails.

### LINKED ISSUE #19564: 🤖 Add retry button in the all domains error screen
Add a button in the all domains error screen so that the user can retry fetching without leaving the screen.
Ref: https://github.com/wordpress-mobile/WordPress-Android/pull/19555#pullrequestreview-1718539017

### PR DESCRIPTION
**Pull request by mkevins**
Fixes #19562

### Description

This PR adds a "pull-to-refresh" behavior to the all domains error screen. Also, it updates the error messaging to make it more clear / discoverable.

To test:

1. Enable the `domain_management` feature flag
2. Set the device on flight mode
3. Open the domain management (`Me>Domains` or `My Site>More>Domains>ALL`)
4. Notice that you see the error screen
5. Pull to refresh
6. Notice that nothing happens
7. Disable flight mode
8. Pull to refresh
9. Verify that the domains are fetched

## Regression Notes
1. Potential unintended areas of impact
Domain Management feature

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated tests

3. What automated tests I added (or what prevented me from doing so)
A unit tests was added for the `onRefresh` function

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainManagementActivity.kt [hunk 1]
```diff
@@ -10,6 +10,7 @@ import dagger.hilt.android.AndroidEntryPoint
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
 import org.wordpress.android.ui.ActivityLauncher
+import org.wordpress.android.ui.domains.management.DomainManagementViewModel.*
```
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainManagementActivity.kt]: Adding a minor suggestion fixing the `WildcardImport` Detekt issue

```suggestion
import org.wordpress.android.ui.domains.management.DomainManagementViewModel.ActionEvent
```

**Review by antonis — COMMENTED** [review]
Great work @mkevins 🙇 
Though the PR is in draft state everything worked as expected for me on a Pixel 5 (Android 14) with different configurations. The code changes also look good 🎉 

If that helps I used the following main steps for testing:

1. Enable the `domain_management` feature flag
2. Set the device on flight mode
3. Open the domain management (`Me>Domains` or `My Site>More>Domains>ALL)
4. Notice that you see the error screen
5. Pull to refresh
6. Notice that nothing happens
7. Disable flight mode
8. Pull to refresh
9. **Verify** that the domains are fetched

|Normal|Night|Landscape|
|---|---|---|
|Screenshot_20231108_131038|Screenshot_20231108_131116|Screenshot_20231108_131126|

Nitpick:
We could change the messaging to make the functionality more discoverable. E.g:
```
    <string name="domain_management_error_title">Couldn\'t retrieve your domains</string>
    <string name="domain_management_error_subtitle">Check that you\'re online and pull to refresh.</string>
```

**Comment by justtwago** [comment]
Tested and can confirm that it works as expected! 
I have only one minor suggestion. What if we displayed a pull-to-refresh loader until get a result from API? At this moment, when we pull to refresh, the loader immediately disappears and is replaced by the blinking stripes. The UX looks quite less smooth because of that.

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainManagementActivity.kt (hunk 1 above)
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/domains/management/DomainManagementActivity.kt]: How did that sneak in there? Thank you!

**Comment by mkevins** [comment]
> What if we displayed a pull-to-refresh loader until get a result from API? At this moment, when we pull to refresh, the loader immediately disappears and is replaced by the blinking stripes. The UX looks quite less smooth because of that.

This is a good point. I was on the fence about it. I had it this way at one point, but then changed it back to the stripes for consistency and simplicity. Then, the pull-to-refresh animation before releasing (while the finger is still pressed) did not include the spinner, so it felt less discoverable, and I added that back.

That said, I agree this is not ideal either way 🤷‍♂️ . The drawbacks of using only the spinner is that we'll have introduced _two_ "in progress" indicator animations for the same request, and also it's slightly more complex: we have to wire the state through the tree, introduce another ui state object / class, and handle those state changes where it is hoisted.

I wonder whether an alternative would be to keep the logic as is, and deal with the smoothness with a transition. 🤔

File: WordPress/src/test/java/org/wordpress/android/ui/domains/management/DomainManagementViewModelTest.kt [hunk 2]
```diff
@@ -117,6 +125,23 @@ class DomainManagementViewModelTest : BaseUnitTest() {
         }
     }
 
+    @Test
+    fun `onRefresh fetches all the domains again`() = test {
```
Comment (antonis) [inline WordPress/src/test/java/org/wordpress/android/ui/domains/management/DomainManagementViewModelTest.kt:129]: Thank you for the added test case Matthew 🙇

**Review by antonis — APPROVED** [review]
Awesome work @mkevins 🏅 
I retested the implementation with the latest changes and worked as expected for me. The code also looks good 🎉 
I will proceed with merging the PR and we can follow up with an enhancement PR to tackle [the loading ui improvements](https://github.com/wordpress-mobile/WordPress-Android/pull/19565#issuecomment-1803000480) if needed.

**Comment by mkevins** [comment]
Thank you for reviewing and testing this!

**Pull request merged**